File: Leetcode Challenge/March/Python/Linked List Cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if head == None or head.next == None:
            return False;

        slow  = head;
        fast  = head;

        while fast.next != None and fast.next.next != None :
            slow = slow.next;
            fast = fast.next.next;
            if slow == fast :                    
                return True;
            
        
        return False

File: Leetcode Challenge/March/Java/Linked List Cycle.cpp
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
        if (head == null || head.next == null)
            return false;

        ListNode slow  = head;
        ListNode fast  = head;

        while (fast.next != null && fast.next.next != null) {
            slow = slow.next;
            fast = fast.next.next;
            if (slow == fast) {                      
                return true;
            }
        }
        return false; 
    }
}
